src/Sylius/Bundle/CoreBundle/Form/EventSubscriber/AvailableShippingMethodsSubscriber.php

```diff
@@ -0,0 +1,82 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Bundle\CoreBundle\Form\EventSubscriber;
+
+use Sylius\Component\Addressing\Matcher\ZoneMatcherInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+class AvailableShippingMethodsSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var ZoneMatcherInterface
+     */
+    private $zoneMatcher;
+
+    /**
+     * @param ZoneMatcherInterface $zoneMatcher
+     */
+    public function __construct(ZoneMatcherInterface $zoneMatcher)
+    {
+        $this->zoneMatcher = $zoneMatcher;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            FormEvents::PRE_SET_DATA => 'preSetData'
+        ];
+    }
+
+    /**
+     * @param FormEvent $event
+     */
+    public function preSetData(FormEvent $event)
+    {
+        $data = $event->getData();
+        $form = $event->getForm();
+
+        $form->add('shipments', 'collection', [
+            'type' => 'sylius_checkout_shipment',
+            'label' => false,
+            'options' => [
+                'criteria' => [
+                    'enabled' => true,
+                    'zone' => $this->getZonesForOrder($data),
```

Comment:
Not exactly:

Here we should only have ``collection`` type using ``sylius_shipment`` (we can most likely remove the ``sylius_checkout_shipment`` type). And ``sylius_shipment`` type will add a ``sylius_shipping_method_choice`` type with option ``subject`` equal to currently edited shipment - so it has to be done via listener.